### PR TITLE
Add events for balance reserve and unreserve functions

### PR DESCRIPTION
**Pull request by joepetrowski**
Closes https://github.com/paritytech/substrate/issues/6329

Two things in particular I could use feedback on:
1. Are all the `clone`s OK? It didn't compile without them (or using references), but maybe there's a more elegant solution.
2. I put the `derive(...)` impls on `BalanceStatus` that copied other enums that emit in events. Again, maybe another way that's better.

### CODE REVIEW

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -1150,6 +1157,7 @@ impl<T: Trait<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I>
 		Self::try_mutate_account(who, |account, _| -> DispatchResult {
 			account.free = account.free.checked_sub(&value).ok_or(Error::<T, I>::InsufficientBalance)?;
 			account.reserved = account.reserved.checked_add(&value).ok_or(Error::<T, I>::Overflow)?;
+			Self::deposit_event(RawEvent::Reserved(who.clone(), value.clone()));
```
Comment (bkchr) [inline frame/balances/src/lib.rs:1160]: These deposit events mean we need to change the weight formulas?

Comment (joepetrowski) [reply frame/balances/src/lib.rs:1160]: Possibly, @shawntabrizi ?

Comment (shawntabrizi) [reply frame/balances/src/lib.rs:1160]: No, events are free for any particular dispatch.

We already know that this storage item will be read from and written to once per block, and the new events append api keeps the cost minimal.

File: frame/balances/src/lib.rs [hunk 2]
```diff
@@ -1166,6 +1174,7 @@ impl<T: Trait<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I>
 			// defensive only: this can never fail since total issuance which is at least free+reserved
 			// fits into the same data type.
 			account.free = account.free.saturating_add(actual);
+			Self::deposit_event(RawEvent::Unreserved(who.clone(), actual.clone()));
```
Comment (kianenigma) [inline frame/balances/src/lib.rs:1177]: these vars like `who` can be moved, do we need clone?

File: frame/balances/src/tests.rs [hunk 3]
```diff
@@ -168,7 +168,7 @@ macro_rules! decl_tests {
 						<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath),
 						Error::<$test, _>::LiquidityRestrictions
 					);
-					assert_noop!(
+					assert_err!(
 						<Balances as ReservableCurrency<_>>::reserve(&1, 1),
 						Error::<$test, _>::LiquidityRestrictions
```
Comment (kianenigma) [inline frame/balances/src/tests.rs:173]:
```suggestion
						Error::<$test, _>::LiquidityRestrictions,
```

File: frame/balances/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply frame/balances/src/lib.rs:1177]: Only those that are still used in scope (`actual`) need to be cloned.

File: frame/balances/src/tests.rs [hunk 4]
```diff
@@ -168,9 +168,9 @@ macro_rules! decl_tests {
 						<Balances as Currency<_>>::transfer(&1, &2, 1, AllowDeath),
 						Error::<$test, _>::LiquidityRestrictions
 					);
-					assert_noop!(
+					assert_err!(
```
Comment (shawntabrizi) [inline frame/balances/src/tests.rs:171]: Why did you change this from noop to err?

Comment (shawntabrizi) [reply frame/balances/src/tests.rs:171]: Im guessing because we are emitting an event and changing the state? I think this violates a rule that errors shouldnt change state?

File: frame/balances/src/lib.rs (hunk 2 above)
Comment (gavofyork) [inline frame/balances/src/lib.rs:1177]: should be moved to the bottom.

File: frame/balances/src/lib.rs (hunk 1 above)
Comment (gavofyork) [inline frame/balances/src/lib.rs:1160]: should be moved to the bottom.

File: frame/balances/src/lib.rs [hunk 5]
```diff
@@ -1217,6 +1226,7 @@ impl<T: Trait<I>, I: Instance> ReservableCurrency<T::AccountId> for Module<T, I>
 					Status::Reserved => to_account.reserved = to_account.reserved.checked_add(&actual).ok_or(Error::<T, I>::Overflow)?,
 				}
 				from_account.reserved -= actual;
+				Self::deposit_event(RawEvent::ReserveRepatriated(slashed.clone(), beneficiary.clone(), actual.clone(), status));
```
Comment (gavofyork) [inline frame/balances/src/lib.rs:1229]: just move it to the bottom. then you know it'll only happen if the operation is successful.

File: frame/balances/src/tests.rs [hunk 6]
```diff
@@ -683,6 +694,48 @@ macro_rules! decl_tests {
...
+					let _ = Balances::deposit_creating(&1, 100);
+
+					System::set_block_number(2);
+					let _ = Balances::reserve(&1, 10);
+
+					assert_eq!(
+						events(),
+						[
+							Event::system(system::RawEvent::NewAccount(1)),
+							Event::balances(RawEvent::Endowed(1, 100)),
```
Comment (shawntabrizi) [inline frame/balances/src/tests.rs:711]: I would suggest modifying the test here to not check for other events that are not within the scope of this test. Will help prevent other future changes from breaking this test.

So i would index into the last element of the vec, and simply check just that one item matches reserved
